### DISCUSSION TITLE
Mullvad+Shadowsocks Beginner's Questions

### DISCUSSION THREAD

**Discussion by MorphogeneticFields** [qdm12/gluetun#1261] [Q&A]
I am kind of new to docker, but here it goes. 

I'd like to deploy a gluetun container with Mullvad VPN (wireguard) and shadowsocks running. I'm already stuck with the docker config, which returns this error: _docker run requires at least 1 argument_ 

Here is my draft config:

` docker run -it --rm \
--cap-add=NET_ADMIN \
--name=gluetun \
-e VPN_SERVICE_PROVIDER=mullvad \
-p 8888:8888/tcp # HTTP proxy \
-p 8388:8388/tcp # Shadowsocks \
-p 8388:8388/udp # Shadowsocks \
-v /apps/docker/gluetun:/gluetun \
-e VPN_TYPE=wireguard \
-e WIREGUARD_PRIVATE_KEY=xxxx= \
-e WIREGUARD_ADDRESSES="xxx" \
-e SHADOWSOCKS=on \
-e SERVER_CITIES=xxx \
qmcgaw/gluetun
`

The goal is to point my host's browser to the shadowsocks server through an add on such as FoxyProxy so that you can toggle between accessing the internet directly or through the proxy. 

Would that work with my config above after eliminating the error it is currently returning?
Further, I'd like to know if no Mullvad config file has to be put anywhere?
What steps would be needed to connect FoxyProxy to shadowsocks?
How would I find out the gluetun container's local IP address?

Thanks in advance, if anyone can point me in the right direction.

**Comment by qdm12** [comment]
Welcome to Docker and Gluetun 🎉

> I'm already stuck with the docker config, which returns this error: docker run requires at least 1 argument

Your command looks fine, maybe try putting it all on one line without the `\` escapes and removing the comments? The error is usually when you forget the image name (here qmcgaw/gluetun) at the end, or your shell thinks it's 2 distinct commands instead of one. Also make sure you use `"` for values containing spaces such as `SERVER_CITIES` eventually.

```sh
docker run -it --rm --cap-add=NET_ADMIN --name=gluetun -e VPN_SERVICE_PROVIDER=mullvad -p 8888:8888/tcp -p 8388:8388/tcp -p 8388:8388/udp -v /apps/docker/gluetun:/gluetun -e VPN_TYPE=wireguard -e WIREGUARD_PRIVATE_KEY=xxxx -e WIREGUARD_ADDRESSES="xxx" -e SHADOWSOCKS=on -e SERVER_CITIES=xxx qmcgaw/gluetun
```

Note you can also run it with `-d` instead of `-it` so it runs in the background and you can then `docker logs -f gluetun`, `docker stop gluetun`.

> Would that work with my config above after eliminating the error it is currently returning?

Yes if foxyproxy supports shadowsocks, I don't totally recall. You might also want to set `SHADOWSOCKS_PASSWORD=xxx` which defaults to an empty string I believe.

> Further, I'd like to know if no Mullvad config file has to be put anywhere?

Nope 🎉 Most of the config values are the same for all users (endpoint ip, port, public address) so you only need to set your private key, address and preshared key if you have set one.

> What steps would be needed to connect FoxyProxy to shadowsocks?

Just set the endpoint to your server lan ip address where gluetun is running, with port 8388. And set the password obviously 😉

> How would I find out the gluetun container's local IP address?

Gluetun has an ip address in the docker network (usually 172.xx.xx.xx), but you don't have to care since you use port mapping `-p 8388:8388` so other devices can just connect through yourServerLanIp:8388 for example `192.168.1.2`.

**Comment by MorphogeneticFields** [comment]
Many thanks for the quick feedback. Really appreciate it! :)  

You were right, removing the comments and some unneeded spaces did the trick. 

However, the container log gives this error message constantly: 

> 2022-11-29T21:24:20Z ERROR [shadowsocks] cannot obtain target address: unexpected EOF

I have pointed my foxyproxy to the LAN IP of the computer running the container and the port 8388. 
Additionally, 
> Send DNS through SOCKS5 proxy

has been activated.

However, when activating the connection in foxyproxy there is just no browsing possible at all, but I guess this makes sense, since shadowsocks is not working properly.

I feel we are close, though. ;)  Any further advice?

**Reply by qdm12** [reply]
Does it say `healthy!` in the logs anywhere? `cannot obtain target address: unexpected EOF` means it cannot DNS resolve or network doesn't work, usually that's because the vpn connection doesn't work. Maybe you can share more of your logs?

**Reply by MorphogeneticFields** [reply]
Yes, Portainer shows the container being healthy. I have pasted a longer log file below. The way I read it, the VPN connection is working. 

I found an article on the Mullvad page: https://mullvad.net/de/help/intro-shadowsocks/
Could it be that there are extra steps needed for Mullvad to work with shadowsocks? 

```
|       ├── Interface addresses:
|       |   └── 10.67.68.169/32
|       └── Network interface: tun0
├── DNS settings:
|   ├── DNS server address to use: 127.0.0.1
|   ├── Keep existing nameserver(s): no
|   └── DNS over TLS settings:
|       ├── Enabled: yes
|       ├── Update period: every 24h0m0s
|       ├── Unbound settings:
|       |   ├── Authoritative servers:
|       |   |   └── cloudflare
|       |   ├── Caching: yes
|       |   ├── IPv6: no
|       |   ├── Verbosity level: 1
|       |   ├── Verbosity details level: 0
|       |   ├── Validation log level: 0
|       |   ├── System user: root
|       |   └── Allowed networks:
|       |       ├── 0.0.0.0/0
|       |       └── ::/0
|       └── DNS filtering settings:
|           ├── Block malicious: yes
|           ├── Block ads: no
|           ├── Block surveillance: no
|           └── Blocked IP networks:
|               ├── 127.0.0.1/8
|               ├── 10.0.0.0/8
|               ├── 172.16.0.0/12
|               ├── 192.168.0.0/16
|               ├── 169.254.0.0/16
|               ├── ::1/128
|               ├── fc00::/7
|               ├── fe80::/10
|               ├── ::ffff:7f00:1/104
|               ├── ::ffff:a00:0/104
|               ├── ::ffff:a9fe:0/112
|               ├── ::ffff:ac10:0/108
|               └── ::ffff:c0a8:0/112
├── Firewall settings:
|   └── Enabled: yes
├── Log settings:
|   └── Log level: INFO
├── Health settings:
|   ├── Server listening address: 127.0.0.1:9999
|   ├── Target address: cloudflare.com:443
|   ├── Read header timeout: 100ms
|   ├── Read timeout: 500ms
|   └── VPN wait durations:
|       ├── Initial duration: 6s
|       └── Additional duration: 5s
├── Shadowsocks server settings:
|   ├── Enabled: yes
|   ├── Listening address: :8388
|   ├── Cipher: chacha20-ietf-poly1305
|   ├── Password: [set]
|   └── Log addresses: no
├── HTTP proxy settings:
|   └── Enabled: no
├── Control server settings:
|   ├── Listening address: :8000
|   └── Logging: yes
├── OS Alpine settings:
|   ├── Process UID: 1000
|   └── Process GID: 1000
├── Public IP settings:
|   ├── Fetching: every 12h0m0s
|   └── IP file path: /tmp/gluetun/ip
└── Version settings:
    └── Enabled: yes
2022-11-30T14:19:16Z INFO IPv6 is not supported
2022-11-30T14:19:16Z INFO [routing] default route found: interface eth0, gateway 172.17.0.1 and assigned IP 172.17.0.3
2022-11-30T14:19:16Z INFO [routing] adding route for 0.0.0.0/0
2022-11-30T14:19:16Z INFO [firewall] setting allowed subnets...
2022-11-30T14:19:16Z INFO [routing] default route found: interface eth0, gateway 172.17.0.1 and assigned IP 172.17.0.3
2022-11-30T14:19:16Z INFO TUN device is not available: open /dev/net/tun: no such file or directory; creating it...
2022-11-30T14:19:16Z INFO [dns over tls] using plaintext DNS at address 1.1.1.1
2022-11-30T14:19:16Z INFO [http server] http server listening on [::]:8000
2022-11-30T14:19:16Z INFO [firewall] allowing VPN connection...
2022-11-30T14:19:16Z INFO [healthcheck] listening on 127.0.0.1:9999
2022-11-30T14:19:16Z INFO [shadowsocks] listening TCP on :8388
2022-11-30T14:19:16Z INFO [shadowsocks] listening UDP on :8388
2022-11-30T14:19:16Z INFO [wireguard] Using available kernelspace implementation
2022-11-30T14:19:16Z INFO [wireguard] Connecting to 185.254.75.4:51820
2022-11-30T14:19:16Z INFO [wireguard] Wireguard is up
2022-11-30T14:19:16Z INFO [dns over tls] downloading DNS over TLS cryptographic files
2022-11-30T14:19:17Z INFO [healthcheck] healthy!
2022-11-30T14:19:19Z INFO [dns over tls] downloading hostnames and IP block lists
2022-11-30T14:19:23Z INFO [dns over tls] init module 0: validator
2022-11-30T14:19:23Z INFO [dns over tls] init module 1: iterator
2022-11-30T14:19:23Z INFO [dns over tls] start of service (unbound 1.15.0).
2022-11-30T14:19:23Z INFO [dns over tls] generate keytag query _ta-4a5c-4f66. NULL IN
2022-11-30T14:19:23Z INFO [dns over tls] generate keytag query _ta-4a5c-4f66. NULL IN
2022-11-30T14:19:24Z INFO [dns over tls] ready
2022-11-30T14:19:24Z INFO [vpn] You are running on the bleeding edge of latest!
2022-11-30T14:19:24Z INFO [ip getter] Public IP address is 185.254.75.40 (Germany, North Rhine-Westphalia, Düsseldorf)
2022-11-30T14:23:37Z ERROR [shadowsocks] cannot obtain target address: unexpected EOF
2022-11-30T14:23:37Z ERROR [shadowsocks] cannot obtain target address: unexpected EOF
2022-11-30T14:23:37Z ERROR [shadowsocks] cannot obtain target address: unexpected EOF
2022-11-30T14:23:37Z ERROR [shadowsocks] cannot obtain target address: unexpected EOF
```

**Reply by qdm12** [reply]
Yeah VPN is working, that's weird :thinking: 
Maybe try with another shadowsocks client?
[In the code](https://github.com/qdm12/ss-server/blob/ecc7356d2b768bc8251f17867bd4251f794ac5f2/pkg/tcp/server.go#L92), that comes from the TCP server where the packet sent from the client is empty (hence EOF = end of file). It may be the client sending wrong packets :thinking: ?

**Reply by MorphogeneticFields** [reply]
I found an up-to-date client for Linux. https://snapcraft.io/shadowsocks-electron
Connecting to the server works, but the error in the container remains. 
One thing I noticed, in the client you have to specify the server type. It can either be "ss or "ssr" (I don't know the difference tbh). ss returns the error from the log while ssr causes  the following:  

ERROR [shadowsocks] cannot obtain target address: chacha20poly1305: message authentication failed.

So it seems the ssr setting at least makes the client send a message, but -alas- it cannot get authenticated. Since I have not set a custom cipher, I don't know why this is.

Fyi, additional required settings in the client are "protocol" (set to "origin") and "obfs" (set to "plain"). 

One question on the side, the client if/when working would reroute the entire system's traffic, right?

Thank you so much for taking the time and seeing me through this!

**Reply by qdm12** [reply]
> It can either be "ss or "ssr" (I don't know the difference tbh). ss returns the error from the log while ssr causes the following:

Yeah ssr is some old no longer maintained version of shadowsocks. With `ss` message encryption/authentication works fine, it's just the message is empty for some reason.

> Since I have not set a custom cipher

Try with chacha20-ietf-poly1305, that's the default in the Gluetun ss server.

> Fyi, additional required settings in the client are "protocol" (set to "origin") and "obfs" (set to "plain").

- protocol: what are the options?
- obfs is probably obfuscation I guess it should be plain

I have been using shadowsocks-qt5 for linux just fine with Gluetun, and I connect switchyOmega (Chrome extension) to connect to it over socks5 on a local port.

> One question on the side, the client if/when working would reroute the entire system's traffic, right?

On Linux, the qt5 program I am using listens on a local socks5 port, and so you can configure your browser to connect to it (which forwards through shadowsocks to Gluetun). Maybe you can forward the entire system network through socks5 but I haven't looked into it.

On Windows, the shadowsocks UI client does magically forward everything (even UDP), same with Android.

**Reply by MorphogeneticFields** [reply]
> Try with chacha20-ietf-poly1305, that's the default in the Gluetun ss server.

That I have already done. I meant that I didn't deploy the container with a custom cypher so that chacha20-ietf-poly1305 should work. 

>     * protocol: what are the options?
 
The protocol options only appear when selecting ssr, and I take it that ss should be used instead.

> I have been using shadowsocks-qt5 for linux just fine with Gluetun, and I connect switchyOmega (Chrome extension) to connect to it over socks5 on a local port.

How do you do connect the browser extension to shadowsocks-qt5, a program I have installed as well. It is connected to the shadowsocks server, but I may be missing the connection to the browser.  
> 
> On Linux, the qt5 program I am using listens on a local socks5 port, and so you can configure your browser to connect to it (which forwards through shadowsocks to Gluetun). Maybe you can forward the entire system network through socks5 but I haven't looked into it.

Actually, it would be convenient if only the browser traffic were to be forwarded.
Shadowsocks-qt5 shows a local port 1080. Pointing the browser to a proxy located at LAN IP:1080 does not work.

**Comment by bnhf** [comment]
@MorphogeneticFields

I've played around with this myself, a couple of times now, and I believe the situation is this:

The easy-to-use browser-extension-based Socks5 proxies don't support ciphers, and the system-wide Socks5 clients don't include a way to route DNS requests via the proxy.

For situations where people want to be able to quickly toggle a proxy on and off, and Gluetun is running on their LAN with Shadowsocks enabled, additional encryption seems unnecessary. In this use-case, it would be ideal to be able to disable encryption and password protection through environment variables. @qdm12 would this be possible to implement?

In the meantime, the best workaround I'm aware of, is to put a Firefox container into your Gluetun stack. It's convenient, and relatively foolproof. It's also nice for confirming details of your VPN connection using websites like ipleak.net, speedtest.net and the various webpages of your VPN provider itself.

Here's an example of a stack with a Firefox container accessible at port 3000:

```yml
version: '3.7'
services:
    gluetun:
        image: qmcgaw/gluetun:latest
        container_name: gluetun
        cap_add:
            - NET_ADMIN
        ports:
            - 8100:8000 # Remote Control VPN
            - 8388:8388 # Shadowsocks
            - 8388:8388/udp # Shadowsocks
            - 3000:3000 # Firefox
        environment:
            - VPN_SERVICE_PROVIDER=airvpn
            - VPN_TYPE=wireguard
            - SERVER_COUNTRIES=Canada          
            - WIREGUARD_PRIVATE_KEY=${WG_PRIVATE}
            - WIREGUARD_PRESHARED_KEY=${WG_PRESHARED}
            - WIREGUARD_ADDRESSES=${WG_ADDRESSES}
            - SHADOWSOCKS=ON
            - SHADOWSOCKS_CIPHER=aes-128-gcm
            - SHADOWSOCKS_PASSWORD=${SS_PWD}
            - PUID=1000
            - PGID=1000
            - TZ=${TZ}
        volumes:
            - /data/wireguard:/gluetun
  
    firefox:
        image: lscr.io/linuxserver/firefox:latest
        container_name: firefox
        environment:
            - PUID=1000
            - PGID=1000
            - TZ={$TZ}
        volumes:
            - /data/firefox:/config
        shm_size: '1gb'
        network_mode: 'service:gluetun'
        depends_on:
            - gluetun
```
Variables in shown in "${ }" would be added added the bottom of a Portainer stack in the Environment section without the "${ }"

**Reply by MorphogeneticFields** [reply]
That sounds like a great solution for the time being. Never thought about running a browser in a container. The Firefox in the container would have to be accessed through another browser on the host side, then? I am curious how that would look like.

Since I haven't used stacks so far, one question regarding your last sentence. If my shadowsocks password were "password" without the quotes, the end of the environment section would have to look like this? 

```
 - TZ=${TZ}
 - SHADOWSOCKS_PASSWORD=password
```

**Reply by bnhf** [reply]
@MorphogeneticFields 

You can just replace an environment variable with the actual value if like, such as putting US/Mountain in place of "${TZ}".

Or if you want to utilize the variable capability of Portainer, you can keep the variables in the stack , and assign them values in the environment section of the Portainer stack you're creating like so:

![Screenshot 2022-12-01 075636](https://user-images.githubusercontent.com/41088895/205087031-bfe76dca-2448-4e32-a8e3-9a0634ed27b9.png)

And if you haven't tried it yet -- yes, Firefox in a container is accessed using another browser on your network. Any browsing you do in that "nested" browser window will go through your VPN. It's quite slick for a variety of purposes.

**Reply by qdm12** [reply]
Yeah I did that as well for some time, but it uses a bit too much resources to my taste :smile_cat: 

If your client is a Linux laptop/desktop then yeah I don't really know how to forward UDP traffic (DNS) through the socks5 listening port created by the shadowsocks client app.

But for Android clients and Windows client, Shadowsocks tunnels UDP/DNS traffic as well so it's worth it, even outside your LAN :wink:

**Reply by MorphogeneticFields** [reply]
@bnhf 
Would you mind explaining in layman terms what WG_PRESHARED and WG_ADDRESSES do and what values` to add there ? 

When downloading a wireguard config file, it contains a private key, a public key and an address. 

Could you create two wireguard files, enter two sets of private keys and addresses, and gluetun would randomly switch between the VPN servers upon start or would a second address only work as a fallback?

**Reply by bnhf** [reply]
@MorphogeneticFields 

Apologies if I muddied the waters by posting this stack using environment variables. It's not necessary to use them -- they can be replaced right in the stack with the actual values if you like.

As to what these values are, a Wireguard pre-shared key (which I represented as the environment variable WG_PRESHARED) is an optional improvement to encryption that may or may not be used by your provider. If Mullvad didn't supply one, you can comment out that line using a "#". The Wireguard addresses (which I represented as WG_ADDRESSES) are the IPV4 and IPV6 addresses assigned to you by your VPN provider on their private network. They will only accept data from you originating from these addresses.

Copy the private key from the Mullvad Wireguard config file and either use it to replace "${WG_PRIVATE}" in the stack, or keep the variable and use it in the environment section as the "value" opposite WG_PRIVATE. Same story for the address or addresses they provide, which would replace "${WG_ADDRESSES}", or be used as the "value" if you want to keep the environment variable approach. Comment out the line in the stack with ```WIREGUARD_PRESHARED_KEY=${WG_PRESHARED}``` since Mullvad didn't provide one.

**Reply by MorphogeneticFields** [reply]
No need for apologies. You explained it just fine. In fact, everything seems to be working. Thanks a lot.

Just some observations after the first few minutes of playing around with the containerized Firefox. 

Is it possible to share the host's clipboard with Firefox?
What is the below setting for? Firefox seems to be a bit laggier compared to the host version. Is this setting for RAM?
`shm_size: '1gb'`

The Mullvad check shows DNS leaks for the containerized Firefox but the "leak" is that some Cloudflare servers from the Netherlands are shown, so it is not really a leak because these are not my ISP's DNS servers. Then again, these Cloudflare servers do not appear when connecting the host directly with the Mullvad app. Nevertheless, it is not really a leak, imo.

**Comment by bnhf** [comment]
@MorphogeneticFields 

I found a Socks5 server container that doesn't appear to require a cipher or password. Here's the compose to try in Portainer Stacks:

```yml
version: '3.7'
services:
    gluetun:
        image: qmcgaw/gluetun:latest
        container_name: gluetun
        cap_add:
            - NET_ADMIN
        ports:
            - 8100:8000 # Remote Control VPN
            - 8388:8388 # Socks5
            - 8388:8388/udp # Socks5
            - 3000:3000 # Firefox
        environment:
            - VPN_SERVICE_PROVIDER=airvpn
            - VPN_TYPE=wireguard
            - SERVER_COUNTRIES=Canada          
            - WIREGUARD_PRIVATE_KEY=${WG_PRIVATE}
            - WIREGUARD_PRESHARED_KEY=${WG_PRESHARED}
            - WIREGUARD_ADDRESSES=${WG_ADDRESSES}
            - PUID=1000
            - PGID=1000
            - TZ=${TZ}
        volumes:
            - /data/wireguard:/gluetun
            
    socks5:
        image: serjs/go-socks5-proxy:latest
        container_name: socks5
        environment:
            - PROXY_PORT=8388
            - TZ=${TZ}
        network_mode: 'service:gluetun'
        depends_on:
            - gluetun
              
    firefox:
        image: lscr.io/linuxserver/firefox:latest
        container_name: firefox
        environment:
            - PUID=1000
            - PGID=1000
            - TZ={$TZ}
        volumes:
            - /data/firefox:/config
        shm_size: '1gb'
        network_mode: 'service:gluetun'
        depends_on:
            - gluetun
```
Obviously it would be nice to have this implemented directly in Gluetun, but in the interim, this stack should give you the ability to use Foxy Proxy or perhaps Proxy Toggle (which seemed even more convenient to me).

@qdm12 I'm guessing from the container name that this is written in Go, so perhaps there are some clues to add a no-cipher option to Gluetun. :-)

**Reply by MorphogeneticFields** [reply]
Another great idea, I guess any SOCKS5 proxy would do for me. It doesn't have to be shadowsocks. 
Do all Socks5 proxies obfuscate the fact that one is behind a VPN? So, could I try to log into my bank account without fearing to get the account locked?

I will try to use your stacks. Thanks!

**Reply by bnhf** [reply]
Any detection of whether you're using a VPN, as far as I know, is based on your exit IP. So, using Socks5/Shadowsocks in the way we're using it, does nothing to change that exit node. All we're doing here is setting up an easy way to send traffic over your container-based VPN from any browser on your LAN.

**Reply by bnhf** [reply]
@MorphogeneticFields 

Are you trying to do something like this?
https://mullvad.net/en/help/socks5-proxy/

**Reply by MorphogeneticFields** [reply]
Yes, this is part of it. It all starts with this: [https://mullvad.net/de/blog/2021/8/25/when-split-tunneling-useful/](url)

There are use cases where a VPN connection may not be the best option. Split tunneling is one solution.

However, if you read through the link, it says on page 2 that split tunneling settings in the Mullvad app are not permanent, on Linux that is. You'd have to renew everything whenever the VPN state changes. The VPN runs on my not always on Linux desktop, so this is no option.

Next step is proxy servers. They have benefits of their own, but I cannot use the Mullvad proxy directly because it requires the VPN to be running all the time.

All of this is to find a workaround. An HTTP proxy would cause DNS leaks, so a socks5 proxy sounds better to me. I think there is a degree of cloaking the VPN connection as well if you look at this quote from Mullvad.

> Another benefit is a reduction in the amount of CAPTCHAs you will experience. Many websites and services require this identification when they detect traffic that originates from a VPN server. The proxy makes this detection more difficult.

Best case scenario for maximum convenience would be that even banking would be possible when browsing behind the proxy behind the VPN with the VPN endpoint set to my country of residence...

:)

**Reply by bnhf** [reply]
@MorphogeneticFields 
> Next step is proxy servers. They have benefits of their own, but I cannot use the Mullvad proxy directly because it requires the VPN to be running all the time.

So Gluetun might be able to do what you're looking to do, but not by utilizing a Socks5/Shadowsocks server on your LAN. Mullvad's approach where the proxy server is running on the VPN server side of the equation (rather than on the VPN client side) is what you're after.

You should be able to utilize the containerized Firefox I mentioned earlier, and configure that Firefox to use the Socks5 proxy that Mullvad describes in the link I sent you previously. Gluetun provides the full-time VPN connection, and Firefox accesses the proxy server over that VPN. You'd then be able to access that container, using whatever browser, from anywhere on your LAN.

**Reply by MorphogeneticFields** [reply]
I like these moments of sudden clarity. You are right, the proxy server running on the VPN server side of the equation is what I am looking for. So thanks again for pointing that out!

Didn't have the time yet, but I will try your stacks approach. Already watched a tutorial and it seems doable.

**Reply by qdm12** [reply]
1. Socks5 server in Gluetun: subscribe to #234 for a socks5 server in Gluetun, although note a SOCKS5 proxy won't tunnel UDP/DNS more than Shadowsocks does (depending on client/host like Shadowsocks). Since Socks5 isn't encrypted, the http proxy would do just fine as a replacement, and you can use shadowsocks for a secured connection between your client and Gluetun. I'll implement it eventually, but it really is not a priority.
2. Split tunneling: I don't see how a socks5 proxy is relevant to split tunneling, and why you would want split tunneling in Gluetun. Split tunneling is useful to VPN only a part of the traffic, but for Gluetun it doesn't make sense, we want all the gluetun traffic vpn'ed as far as I know.
3. Proxy after Wireguard/OpenVPN exit: why would you want that? It just sounds like extra encryption for no added security? Just wireguard directly to the exit node you want and that's it? Or did I miss something?

**Reply by MorphogeneticFields** [reply]
re 2+3. This is mostly a workaround. By connecting to gluetun -through the proxy- I effectively split my system's traffic in that I decide what goes through the VPN and what not. In the middle of trying to accomplish this, I learned that socks5 proxies also have some advantages over a direct VPN connection. This is described in the link above. It is not so much for an extra layer of encryption if there is any, but rather to circumvent website's anti-VPN-measures. At least this is how I understand it.

Another benefit of the "Firefox being in the container as well" solution would be the prevention of DNS leaks. 
But you guys are the experts, and I may be completely wrong. :)

edit: just saw that qdm12 summarized it well below.

**Reply by qdm12** [reply]
Got it thanks!

> edit: just saw that qdm12 summarized it well below.

it was @bnhf (thanks btw), not me 😅

**Comment by bnhf** [comment]
@qdm12 

I believe this discussion started with a misunderstanding, so let me attempt to distill this down:

@MorphogeneticFields wants to take advantage of a specific feature of Mullvad that allows someone connected to their VPN network to access a Socks5 proxy on Mullvad's internal network. For example, if one was actively using Mullvad's VPN client on their PC, they could then configure their browser to access that proxy at ```10.64.0.1:1080``` (Wireguard) or ```10.8.0.1:1080``` (OpenVPN). According to Mullvad this gives you a unique static exit IP, and reduces CAPTCHAs when browsing via the VPN.

The built-in Shadowsocks server in Gluetun is great, but does nothing as far as I can see to help access this Mullvad feature. Maybe I'm the one that's missing something, but that's my understanding of where this discussion rests at the moment. :-)

**Reply by qdm12** [reply]
Ah interesting, my bad for overlooking this. Not too sure why Mullvad wouldn't expose those static IPs directly through wireguard or openvpn though.

Gluetun doesn't support such thing unfortunately (and feels oof to implement, especially if it's only for Mullvad). Isn't it possible to chain socks5 (or http proxy + socks5)? Or #739 would help in that case (but don't have a timeline for it, I'm pretty much out of time all the time)
